src/main/scala/viper/gobra/ast/internal/transform/OverflowChecksTransform.scala

```diff
@@ -0,0 +1,146 @@
+      Method(receiver, name, args, results, pres, posts, body map (computeNewBody(args, _)))(m.info)
+
+    // Adds pre-conditions stating the bounds of each argument and a post-condition to check if the body expression
+    // overflows
+    case f@PureFunction(name, args, results, pres, posts, body) => body match {
+      case Some(expr) =>
+        val newPre = pres ++ getPureBlockPres(args)
+        val newPost = posts ++ getPureBlockPosts(expr, results)
+        PureFunction(name, args, results, newPre, newPost, body)(f.info)
+      case None => f
+    }
+
+    // Same as pure functions
+    case m@PureMethod(receiver, name, args, results, pres, posts, body) => body match {
+      case Some(expr) =>
+        val newPre = pres ++ getPureBlockPres(args)
+        val newPost = posts ++ getPureBlockPosts(expr, results)
+        PureMethod(receiver, name, args, results, newPre, newPost, body)(m.info)
+      case None => m
+    }
+
+    /* As discussed on the Gobra meeting (27/10/2020), overflow checks should not be added to predicates, assertions
+    * and any other purely logical (i.e. non-executable code) statements and expressions. This seems to be the approach taken
+    * by other verification tools such as FramaC, as noted by Wytse
+    */
+
+    case x => x
+  }
+
+  /**
+    * Adds overflow checks to the bodies of a block. If block is parametrized,
+    * `computeNewBody` adds assumptions about the parameters' bounds
+    */
+  private def computeNewBody(args: Vector[Parameter.In], body: Block): Block = {
+    val assumeArgBounds = args map { case p@Parameter.In(_, typ) => Assume(assertionExprInBounds(p, typ)(p.info))(p.info) }
+    val blockStmts = body.stmts map stmtTransform
+    Block(body.decls, assumeArgBounds ++ blockStmts)(body.info)
+  }
+
+  /**
+    * Computes the pre-conditions to be added to pure functions and methods to check for overflows, i.e.
+    * that each argument is within the bounds for its type
+    */
+  private def getPureBlockPres(args: Vector[Parameter.In]): Vector[Assertion] = {
+    args.map { p: Parameter.In => assertionExprInBounds(p, p.typ)(p.info) }
+  }
+
+  /**
+    * Computes the post-conditions to be added to pure functions and methods to check for overflows, i.e.
+    * that the expression result is within the bounds of its type
+    */
+  private def getPureBlockPosts(body: Expr, results: Vector[Parameter.Out]): Vector[Assertion] = {
+    // relies on the current assumption that pure functions and methods must have exactly one result argument
+    if (results.length != 1) violation("Pure functions and methods must have exactly one result argument")
+    Vector(assertionExprInBounds(body, results(0).typ)(addAnnotation(body.info)))
+  }
+
+  private def stmtTransform(stmt: Stmt): Stmt = stmt match {
+    case b@Block(decls, stmts) => Block(decls, stmts map stmtTransform)(b.info)
+
+    case s@Seqn(stmts) => Seqn(stmts map stmtTransform)(s.info)
```

Comment:
I think the rewriting framework here is a bit flawed, but we can change that later. 
At an earlier point in time, Gobra reused Viper's rewriting framework. Unfortunately, due to changes to Viper this was not possible anymore (Before the framework was independend of the Viper AST. Then reflections were added thatenforce that the ast must have the Viper meta data ...). Back then, I objected to this change heavily, because now (still?) we would have to copy the entire rewriting framework and change how the meta data is handled. It might be worth to check if this is still the case and if not, then we should reuse Viper's framework again.

Reply:
Ah, I think that the rewriting framework is flawed because there is a lot of redundant code like the two cases that I marked. If more compound statements are added, then you need to add these lines for each of them.

Reply:
Sorry for the late reply to your comment. Should I bring this up in the next Viper meeting? In the case that we cannot re-use the rewriting framework from Viper, I will improve this (maybe in a future PR).

Reply:
That is a good idea, I might ask this at the meeting if the situation fits.

Reply:
For completeness sake, I'm pasting this message here
> in a first look and assuming that it is implemented in the traits Rewritable (https://github.com/viperproject/silver/blob/master/src/main/scala/viper/silver/ast/utility/rewriter/Rewritable.scala) and Strategy (https://github.com/viperproject/silver/blob/master/src/main/scala/viper/silver/ast/utility/rewriter/Strategy.scala), it seems that the framework is dependent on the Viper AST and I am not sure if we can reuse that in a non-hacky way